cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_utils.go

```diff
@@ -143,3 +152,38 @@ func machineSetIsOwnedByMachineDeployment(machineSet *MachineSet, machineDeploym
 	}
 	return false
 }
+
+func scaleFromZeroEnabled(annotations map[string]string) bool {
+	if val, exists := annotations[nodeGroupScaleFromZero]; exists {
```

Comment:
can we get this implicitly by checking the existence  of the annotations?

Reply:
makes sense, i have a couple questions:

1. must all these annotations be defined? (eg can a resource have cpu and mb but not gpu)
1. should i make these keys package level exports so that the providers can vendor them in? (my concern is these variables becoming out of sync between repos)

Reply:
added a fix for this

Reply:
> can we get this implicitly by checking the existence of the annotations?

I was previously using this as a brute-force switch to test the behaviour.

However, what's unclear is does the presence of some/all/one annotation in the scale-from-zero-set imply that this machine set can scale from zero?

Reply:
We have control over which providers we officially support autoscaling from/to zero as a product.
We signal this to the autoscaler by enabling a controller in any given platform which sets the well known annotations we choose. https://github.com/openshift/cluster-api-provider-aws/blob/master/pkg/actuators/machineset/controller.go#L108-L110
i.e 
```
if has cpuKey && memKey && gpuKey
 return true
```

maxPods should optional and default to 255 here, see last bullet point https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/autoscaling-from-to-zero.md#machine-api-autoscaler-provider

Reply:
I think GPU should potentially be optional. I've started working on the GCP implementation and as far as I can tell, at present MAPI doesn't support adding GPUs for GCP instances. Seems odd for the controller to just permanently set that value to zero?

Reply:
gpu definitely _seems_ like it should be optional, i'll make some adjustments. thanks!